File: src/2025-03-04-03-28-40--deepseek-chat-0304/cpp/exercises/practice/grade-school/grade_school.cpp
#include "grade_school.h"

namespace grade_school {

void school::add(const std::string& name, int grade) {
    // Check if student already exists in any grade
    for (const auto& [grd, students] : students_by_grade) {
        if (students.find(name) != students.end()) {
            throw std::runtime_error("Student already exists in grade " + std::to_string(grd));
        }
    }
    
    // Add student to the specified grade
    students_by_grade[grade].insert(name);
}

std::vector<std::string> school::grade(int grade) const {
    auto it = students_by_grade.find(grade);
    if (it == students_by_grade.end()) {
        return {};
    }
    
    return std::vector<std::string>(it->second.begin(), it->second.end());
}

std::map<int, std::vector<std::string>> school::roster() const {
    std::map<int, std::vector<std::string>> result;
    
    for (const auto& [grade, students] : students_by_grade) {
        result[grade] = std::vector<std::string>(students.begin(), students.end());
    }
    
    return result;
}

}  // namespace grade_school

File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/grade-school/grade_school.py
class School:
    def __init__(self):
        self._roster = {}
        self._added_status = []

    def add_student(self, name, grade):
        # Check if student already exists in any grade
        for students in self._roster.values():
            if name in students:
                self._added_status.append(False)
                return False
                
        # Add student to grade
        if grade not in self._roster:
            self._roster[grade] = []
        self._roster[grade].append(name)
        self._roster[grade].sort()
        self._added_status.append(True)
        return True

    def roster(self):
        result = []
        for grade in sorted(self._roster.keys()):
            result.extend(self._roster[grade])
        return result

    def grade(self, grade_number):
        return self._roster.get(grade_number, []).copy()

    def added(self):
        return self._added_status.copy()
